src/v/storage/compaction_reducers.cc

```diff
@@ -367,106 +367,65 @@ ss::future<> index_rebuilder_reducer::do_index(model::record_batch&& b) {
-            // base_offset().
-            // A corner case where this is not a problem is when the abort
-            // marker is the first entry in the segment.
-            vlog(
-              stlog.warn,
-              "No ongoing aborted tx found for pid {}, batch {}",
-              pid,
-              b.header());
+    // discard any inflight aborted transactions if we encounter an
+    // abort batch.
+    auto is_tx = b.header().attrs.is_transactional();
+    auto is_control = b.header().attrs.is_control();
+    if (is_tx && is_control) {
+        auto batch_type = _stm_mgr->parse_tx_control_batch(b);
+        auto pid = model::producer_identity(
+          b.header().producer_id, b.header().producer_epoch);
+        if (batch_type == model::control_record_type::tx_abort) {
+            _ongoing_aborted_txs.erase(pid);
         }
-        break;
-    }
     }
 }
 
-bool tx_reducer::handle_tx_data_batch(const model::record_batch& b) {
+bool tx_reducer::can_discard_tx_data_batch(const model::record_batch& b) {
+    if (_transactional_stm_type != stm_type::user_topic_transactional) {
+        return false;
+    }
+    auto is_tx = b.header().attrs.is_transactional();
+    auto is_data = b.header().type == model::record_batch_type::raft_data;
+    auto is_control = b.header().attrs.is_control();
     auto pid = model::producer_identity(
       b.header().producer_id, b.header().producer_epoch);
-    auto discard = _ongoing_aborted_txs.contains(pid);
-    if (discard) {
-        _stats._tx_data_batches_discarded++;
-    }
-    return discard;
+    return is_tx && is_data && !is_control
+           && _ongoing_aborted_txs.contains(pid);
 }
 
-bool tx_reducer::handle_non_tx_control_batch(const model::record_batch& b) {
-    auto type = b.header().type;
-    vassert(
-      type == model::record_batch_type::tx_prepare
-        || type == model::record_batch_type::tx_fence,
-      "{} unknown type encountered",
-      type);
-    // Fence batches cannot be discarded because they contain epoch information
-    // from pids that are tracked in the state machine. We key the records with
-    // pid, so the combination of batch_type + pid should always retain the
-    // latest epoch in the indexer_reducer. OTOH prepare batches can be
-    // discarded.
-    bool discard = type == model::record_batch_type::tx_prepare;
-    if (discard) {
-        _stats._non_tx_control_batches_discarded++;
+bool tx_reducer::can_discard_consumer_offsets_batch(
+  const model::record_batch& b) {
+    if (_transactional_stm_type != stm_type::consumer_offsets_transactional) {
+        return false;
     }
-    return discard;
+    // Remove all transaction related batches (including data) because the
+    // committed data has already been rewritten as separate raft_data batches,
+    // so no need to retain originally written group_prepare_tx batches while
+    // the transaction is in progress.
+    return is_compactible_control_batch(b.header().type);
```

Comment:
So simple!

I wonder why we needed a separate batch type - if we know that we are in the consumer offsets topic, can't we just compact all fence batch types? Is this because in theory ordinary user transactions are also possible for the consumer offsets topic?

Reply:
right, this is more of a code structuring problem because we don't have the "topic context" in all places, that is fixable with some (ugly) refactoring but even otherwise there is no reason to use the same batch type for fence in data & consumer offsets especially when every other batch type (data, commit/abort) are different. This change simplifies things because we can now just look at the batch header and say whether it is compactible and all the batch types are neatly grouped by topic (no overlap).